WikiExtension/src/timeSeriesService.js

```diff
@@ -25,24 +26,97 @@ const fetchPageViews = async (title, startDate, endDate, aggregateType) => {
     return await response.json();
 };
 
+/**
+ * Get all the revision objects on a given wikipedia title.
+ *
+ * The api only gives us at most 20 reivision objects per request. However, it does give us a query to
+ * use if there are any older revisions. If older versions of revision exists, we will get
+ * {
+ *  revisions: [...],
+ *  ...
+ *  older: 'https://en.wikipedia.org/w/rest.php/v1/page/Jupiter/history?older_than=1103899458',
+ *  ...
+ * }
+ *
+ * Thus, we keep calling the endpoint with the given query if the older property exists in the
+ * previous request.
+ *
+ * @param {string} title of a wikipedia article
+ * @param {Date} startDate
+ * @param {Date} endDate
+ * @returns An array containing all revision objects
+ */
+const fetchPageRevisions = async (title, startDate, endDate) => {
+    const response = await fetch(`https://en.wikipedia.org/w/rest.php/v1/page/${title}/history`);
+    let curJsonResponse = await response.json();
+
+    let revisions = curJsonResponse.revisions;
+
+    while (curJsonResponse.hasOwnProperty("older")) {
+        const localResponse = await fetch(curJsonResponse.older);
+        curJsonResponse = await localResponse.json();
+
+        const localRevisions = curJsonResponse.revisions;
+
+        // Stop making requests when requesting passes startDate
+        const localLatestTimestamp = new Date(localRevisions[0].timestamp);
+        if (localLatestTimestamp < startDate) break;
+
+        revisions.push(...localRevisions);
+    }
+
+    // Filters revisions that are in between the given timerange
+    return revisions.filter((revisionObject) => {
+        const currentTimestamp = new Date(revisionObject.timestamp);
+        return startDate <= currentTimestamp && currentTimestamp <= endDate;
+    });
+};
+
 /**
  * TODO: Subject to change based on what the graph needs
- * @return formated fetchedPageViews as a list of lists that only contain date object and count
+ * @return formated fetchedPageViews as an array of arrays that only contain date object and count
  */
 const formatPageViews = (fetchedPageViews) => {
     return Array.from(fetchedPageViews.items).map((pageViewObject) => {
         return [formatYYYYMMDDToDate(pageViewObject.timestamp), pageViewObject.views];
     });
 };
 
+/**
+ * TODO: Subject to change based on what the graph needs
+ * @param {AggregateType} aggregateType
+ * @return formated fetchedRevisions as an array of arrays that only contain date object and count
+ */
+const formatPageRevisions = (fetchedRevisions, aggregateType) => {
+    const pageRevisionCountMap = fetchedRevisions
+        .map((revisionObject) => {
+            const date = new Date(revisionObject.timestamp);
+            date.setHours(0, 0, 0, 0); // have date only contains year, month, and day (not time)
+
+            // Set date to 1 to aggregate monthly instead of daily
+            if (aggregateType == AggregateType.MONTHLY) {
+                date.setDate(1);
```

Comment:
Can you please explain why we need this when doing monthly?

Reply:
in case we want to aggregate data on a monthly basis. it's just an option that `pageView` allows, so I thought I should also do it here.